fastjet/internal/delaunay/delaunay.go

```diff
@@ -7,27 +7,354 @@ package delaunay
+		if !ta.Equals(t) && p.inTriangle(ta) == onEdge {
+			found = true
+			t2 = ta
+			break
+		}
+	}
+	if found {
+		pA1 = p1
+		found = false
+		for _, ta := range p2.adjacentTriangles {
+			if ta.Equals(t2) {
+				found = true
+				break
+			}
+		}
+		if found {
+			pA2 = p2
+			pO1 = p3
+		} else {
+			pA2 = p3
+			pO1 = p2
+		}
+	} else {
+		pO1 = p1
+		for _, ta := range p2.adjacentTriangles {
+			if !ta.Equals(t) && p.inTriangle(ta) == onEdge {
+				found = true
+				t2 = ta
+				break
+			}
+		}
+		if !found {
+			panic(fmt.Errorf("delaunay: can't find second triangle with edge to %v and %v on edge", t, p))
+		}
+		pA1 = p2
+		pA2 = p3
+	}
+	switch {
+	case !t2.A.Equals(pA1) && !t2.A.Equals(pA2):
+		pO2 = t2.A
+	case !t2.B.Equals(pA1) && !t2.B.Equals(pA2):
+		pO2 = t2.B
+	case !t2.C.Equals(pA1) && !t2.C.Equals(pA2):
+		pO2 = t2.C
+	default:
+		panic(fmt.Errorf("delaunay: no point in %v that is not adjacent to the edge of %v", t2, p))
+	}
+	// form four new triangles
+	nt1 := NewTriangle(pA1, p, pO1)
+	nt2 := NewTriangle(pA1, p, pO2)
+	nt3 := NewTriangle(pA2, p, pO1)
+	nt4 := NewTriangle(pA2, p, pO2)
+	updated := nt1.add()
+	updtemp := nt2.add()
+	updated = append(updated, updtemp...)
+	updtemp = nt3.add()
+	updated = append(updated, updtemp...)
+	updtemp = nt4.add()
+	updated = append(updated, updtemp...)
+	updtemp = t.remove()
+	updated = append(updated, updtemp...)
+	updtemp = t2.remove()
+	updated = append(updated, updtemp...)
+	t.children = append(t.children, nt1, nt3)
+	t2.children = append(t2.children, nt2, nt4)
+	d.triangles = append(d.triangles, nt1, nt2, nt3, nt4)
+	// change the edges so it is a valid delaunay triangulation
+	updtemp = d.swapDelaunay(nt1, p)
+	updated = append(updated, updtemp...)
+	updtemp = d.swapDelaunay(nt2, p)
+	updated = append(updated, updtemp...)
+	updtemp = d.swapDelaunay(nt3, p)
+	updated = append(updated, updtemp...)
+	updtemp = d.swapDelaunay(nt4, p)
+	updated = append(updated, updtemp...)
+	return updated
+}
+
+// swapDelaunay finds the triangle adjacent to t and opposite to p.
+// Then it checks whether p is in the circumcircle. If p is in the circumcircle
+// that means that the triangle is not a valid delaunay triangle.
+// Therefore the edge in between the two triangles is swapped, creating
+// two new triangles that need to be checked.
+// It returns the points whose nearest neighbors changed during the
+// process.
+func (d *Delaunay) swapDelaunay(t *Triangle, p *Point) []*Point {
+	// find points in the triangle that are not p
+	var p2, p3 *Point
+	switch {
+	case p.Equals(t.A):
+		p2 = t.B
+		p3 = t.C
+	case p.Equals(t.B):
+		p2 = t.C
+		p3 = t.A
+	case p.Equals(t.C):
+		p2 = t.A
+		p3 = t.B
+	default:
+		panic(fmt.Errorf("delaunay: can't find point %v in Triangle %v", p, t))
+	}
+	// find triangle opposite to p
+	var ta *Triangle
+	for _, t1 := range p2.adjacentTriangles {
+		for _, t2 := range p3.adjacentTriangles {
+			if !t1.Equals(t) && t1.Equals(t2) {
+				ta = t1
+				break
```

Comment:
you're just breaking out of the second, inner, for-loop, not out of the outer for-loop:
https://play.golang.org/p/7Z30S2bJGP

use `break loop` to break out of the outer loop:

```go
loop:
for _, t1 := range p2.adjacentTriangles {
    for _, t2 := range p3.adjacentTriangles {
        if !t1.Equals(t) && t1.Equals(t2) {
            ta = t1
            break loop
        }
    }
}
```